commit/src/domain.rs

```diff
@@ -0,0 +1,364 @@
+use alloc::vec::Vec;
+
+use itertools::Itertools;
+use p3_field::{
+    batch_multiplicative_inverse, cyclic_subgroup_coset_known_order, ExtensionField, Field,
+    TwoAdicField,
+};
+use p3_matrix::dense::RowMajorMatrix;
+use p3_matrix::Matrix;
+use p3_util::{log2_ceil_usize, log2_strict_usize};
+
+use crate::two_adic_decompose::decompose;
+
+pub struct LagrangeSelectors<T> {
+    pub is_first_row: T,
+    pub is_last_row: T,
+    pub is_transition: T,
+    pub inv_zeroifier: T,
+}
+
+pub trait PolynomialDomain: Copy {
```

Comment:
Sorry I think I suggested "domain" before, but if `Val` is part of this then I think "polynomial space" seems more fitting in that case.